Allow to accept multiple values for --security-group-ids and show deprecated message for comma seprated

Vasu1105 opened the pull request.

btm commented on spec/unit/ec2_server_create_spec.rb:
```diff
@@ -2113,5 +2112,39 @@
+        end
+      end
+
+      context "when invalid input is given for --security-group-ids from knife.rb" do
```
is there a reason not support this?

btm replied:
couldn't we just check if we were given an array?

Vasu1105 replied:
@btm We are expecting a input from user as 'x,y,z'. If we have to accept the input as an array to then we need to include that also in help description.

btm replied:
I understand that from the CLI, I wouldn't add array support for the CLI. But using an array would be pretty natural from inside the `knife.rb`.

Vasu1105 replied:
@btm If we do that then we are expecting different input from cli and from knife.rb and its no where in documentation so we tried to avoid this confusion for user and expecting same input format from cli and knife.rb

btm replied:
We should still allow a comma separated list in a string in the configuration file, and use the same code to parse it that we use when we take a comma separated list from the command line.

We should also allow an array in the configuration file. It is ruby, so the user is likely to try to use the array. We allow arrays in other places in the configuration files, like `cookbook_path` and `ohai.disabled_plugins`. We don't have to document it in the user documentation (although it wouldn't hurt). I'd expect a test like `it "allows --security-group-ids set from an Array in knife.rb"`.

btm commented:
@Vasu1105 I talked to some chef maintainers, and we want to standardize on passing a CLI avalue multiple times by calling the argument multiple times instead of a comma separated list.

A) multiple values passed to the CLI is done by passing the flag multiple times
B). use an array in the configuration file

This is so we don't have to deal with parsing issues exactly like this.

Implementing A:

1) Add support to pass -g multiple times like in https://github.com/chef/chef/blob/master/lib/chef/knife/bootstrap.rb#L251-L260

2) detect commas, tell people it's deprecated, and to pass -g multiple times instead

Implementing B:

1) Add support for arrays

2) detect a comma separated string, raise a warning telling people to switch to an array.

Vasu1105 commented:
@btm working on required changes. Fixing specs is in progress.

Vasu1105 commented:
@btm I have created gist for ec2_server_create_spec since git is not showing the diff as it exceeding the number of line changes REF gist https://gist.github.com/Vasu1105/aba8bf2d883bfd51dd717c4598aff9ee.
I have used let and replaced most of the instance variable in specs because of which it showing lot of changes.

btm commented on lib/chef/knife/ec2_server_create.rb:
```diff
@@ -67,10 +67,18 @@ class Ec2ServerCreate < Knife
-        :long => "--security-group-ids 'X,Y,Z'",
-        :description => "The security group ids for this server; required when using VPC,Please provide values in format --security-group-ids 'X,Y,Z'",
-        :proc => Proc.new { |security_group_ids| security_group_ids.split(',') }
+        :short => "-g SECURITY_GROUP_IDS",
```
this should probable read `-g SECURITY_GROUP_ID` since it takes a single value.

btm commented on lib/chef/knife/ec2_server_create.rb:
```diff
@@ -67,10 +67,18 @@ class Ec2ServerCreate < Knife
-        :description => "The security group ids for this server; required when using VPC,Please provide values in format --security-group-ids 'X,Y,Z'",
-        :proc => Proc.new { |security_group_ids| security_group_ids.split(',') }
+        :short => "-g SECURITY_GROUP_IDS",
+        :long => "--security-group-ids",
```
I have a feeling we should change this to `--security-group-id` which probably means having a completely separate `--security-group-ids` option that raises an error and tells you to use the new flag and that you should specify it multiple times instead of using a comma now, then we could take `--security-group-ids` out in a year or so.

Vasu1105 replied:
@btm 
1) So should I show message as deprecated for option --security-group-ids for all kinds of input, whether its comma separated or single value ?
2) In knife.rb the option will remain the same as security_group_ids and it will accept array.
3) Will add the Deprecated message in the --security-group-ids description "[DEPRECATED] Use --security-group-id option instead"

mwrock replied:
I think since its deprecated, it should still "work" until it is removed so I'd keep current validation and functionality in tact with just the additional deprecation notice as mentioned above.

Vasu1105 replied:
@mwrock @btm has suggested to raise an error for option --security-group-ids and show user message as deprecated and use new option --security-group-id. I am making changes in code so please let me know what is the final approach that we should take.

mwrock replied:
I think "deprecation" and raising an error are two different states. I think the safer approach here is to deprecate it now and raise an error in a future release. If we say it is deprecated, that means it should continue to work but we make it very clear that its lifetime is limited. @btm please comment if you have different thoughts but I don't think we can say its deprecated and disallow it at the same time.

btm replied:
Yes @mwrock.

@Vasu1105: Keep whatever behavior we've added for `--security-group-ids` but have it print a warning that you should switch to using `--security-group-id` and specify a single group with each use rather than a list, and that the `--security-group-ids` option will be removed in a future release. the `-g` short flag should go with the new `--security-group-id`.

btm commented on lib/chef/knife/ec2_server_create.rb:
```diff
@@ -67,10 +67,18 @@ class Ec2ServerCreate < Knife
+        :description => "The security group ids for this server; required when using VPC. Mulitple values can be provide as -g sg-e985168d -g sg-e858768d",
+        :proc => Proc.new { |security_group_ids|
+          if security_group_ids.split(',').size > 1
+            Chef::Log.error("[DEPRECATED] Comma separated values for --security-group-ids is deprecated. Provide --security-group-ids option multiple times if mulitple values has to be provided. for e.g. --security-group-ids sg-e985168d --security-group-ids sg-e7f06383 --security-group-ids sg-ec1b7e88 .")
```
[DEPRECATED] Providing comma separated values for --security-group-ids is deprecated. Use the --security-group-ids option multiple times when specifying multiple groups for e.g. --security-group-ids sg-e985168d --security-group-ids sg-e7f06383 --security-group-ids sg-ec1b7e88.

Vasu1105 commented:
@mwrock & @btm I have updated the changes and also updated the gist for spec file https://gist.github.com/Vasu1105/aba8bf2d883bfd51dd717c4598aff9ee. Please review

mwrock commented on lib/chef/knife/ec2_server_create.rb:
```diff
@@ -67,10 +67,28 @@ class Ec2ServerCreate < Knife
-        :long => "--security-group-ids 'X,Y,Z'",
-        :description => "The security group ids for this server; required when using VPC,Please provide values in format --security-group-ids 'X,Y,Z'",
-        :proc => Proc.new { |security_group_ids| security_group_ids.split(',') }
+        :long => "--security-group-ids IDS",
```
I think I'd keep `'X,Y,Z'` over `IDS`

mwrock commented on lib/chef/knife/ec2_server_create.rb:
```diff
@@ -852,9 +875,10 @@ def validate!
+        if locate_config_value(:security_group_ids) && locate_config_value(:security_group_ids).class == String
           ui.error("Invalid value type for knife[:security_group_ids] in knife configuration file (i.e knife.rb). Type should be array. e.g - knife[:security_group_ids] = ['sgroup1']")
-          exit 1
+          exit
```
dhould you still be exiting 1 if this is an error?

Vasu1105 replied:
I have updated the code. The specs where failing when I was using exit 1. Found the issue and fixed the specs.

mwrock commented:
:+1:
